src/protocol/mod.rs

```diff
@@ -0,0 +1,31 @@
+mod bpp;
+mod message;
+
+use crate::errors::ProtocolError;
+
+use std::io::Read;
+
+pub trait Encode {
+    // Consume the entity and return it's BE byte representation.
+    fn encode_to_bytes(self) -> Result<Vec<u8>, ProtocolError>;
```

Comment:
if a method consumes a value, shouldn't there be `...(&mut self)`? Or does it return a new value without changing the original data?

Reply:
After calling 'encode_to_bytes', the original instance no longer exists because I used its data. Self is dropped when the scope of this method ends.

Reply:
If we used a pointer: `&mut self` then we could not really move the data directly, because the instance can persist longer than the method call, by using `self` we take ownership of the instance inside of the method and destruct the self after scope ends, we literally consume the object.